Build error using clang-12/gcc-10

Hi @0vercl0k 

When I tried to build `wtf` on Ubuntu 20.04 with `constexpr bool LoggingOn = true;` in `fuzzer_hevd.cc` I got the below error. I see that by default you are keeping the logging off so you did not encounter this issue.

```
[1/3] Building CXX object CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o
FAILED: CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o 
/usr/bin/g++-10  -DFMT_HEADER_ONLY -I../libs/bochscpu-bins/include -I../libs/robin-map/include -I../libs/kdmp-parser/src/lib -I../libs/libfuzzer -I../libs/readerwriterqueue -I../libs/json/single_include -I../libs/BLAKE3/c -I../libs/CLI11/include -I../libs/fmt/include -I../libs/yas/include -O3 -DNDEBUG -flto -fno-fat-lto-objects   -std=gnu++2a -Winvalid-pch -include /home/ashfaq/Shared/wtf/src/build/CMakeFiles/wtf.dir/cmake_pch.hxx -MD -MT CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o -MF CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o.d -o CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o -c ../wtf/fuzzer_hevd.cc
../wtf/fuzzer_hevd.cc: In instantiation of ‘void Hevd::DebugPrint(const char*, const Args_t& ...) [with Args_t = {}]’:
../wtf/fuzzer_hevd.cc:47:41:   required from here
../wtf/fuzzer_hevd.cc:16:15:   in ‘constexpr’ expansion of ‘fmt::v8::basic_format_string<char>(Format)’
../wtf/fuzzer_hevd.cc:16:15: error: ‘Format’ is not a constant expression
   16 |     fmt::print(Format, args...);
      |     ~~~~~~~~~~^~~~~~~~~~~~~~~~~
../wtf/fuzzer_hevd.cc: In instantiation of ‘void Hevd::DebugPrint(const char*, const Args_t& ...) [with Args_t = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
../wtf/fuzzer_hevd.cc:84:44:   required from here
../wtf/fuzzer_hevd.cc:16:15:   in ‘constexpr’ expansion of ‘fmt::v8::basic_format_string<char, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(Format)’
../wtf/fuzzer_hevd.cc:16:15: error: ‘Format’ is not a constant expression
[2/3] Building CXX object CMakeFiles/wtf.dir/libs/libfuzzer/FuzzerMutate.cpp.o
ninja: build stopped: subcommand failed.
```

Oh no! Will check this out soon, thanks!

Cheers

Le mar. 21 déc. 2021 à 23:05, HackSys Team ***@***.***> a
écrit :

> Hi @0vercl0k <https://github.com/0vercl0k>
>
> When I tried to build wtf on Ubuntu 20.04 with constexpr bool LoggingOn =
> true; in fuzzer_hevd.cc I got the below error. I see that by default you
> are keeping the logging off so you did not encounter this issue.
>
> [1/3] Building CXX object CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o
>
> FAILED: CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o
>
> /usr/bin/g++-10  -DFMT_HEADER_ONLY -I../libs/bochscpu-bins/include -I../libs/robin-map/include -I../libs/kdmp-parser/src/lib -I../libs/libfuzzer -I../libs/readerwriterqueue -I../libs/json/single_include -I../libs/BLAKE3/c -I../libs/CLI11/include -I../libs/fmt/include -I../libs/yas/include -O3 -DNDEBUG -flto -fno-fat-lto-objects   -std=gnu++2a -Winvalid-pch -include /home/ashfaq/Shared/wtf/src/build/CMakeFiles/wtf.dir/cmake_pch.hxx -MD -MT CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o -MF CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o.d -o CMakeFiles/wtf.dir/wtf/fuzzer_hevd.cc.o -c ../wtf/fuzzer_hevd.cc
>
> ../wtf/fuzzer_hevd.cc: In instantiation of ‘void Hevd::DebugPrint(const char*, const Args_t& ...) [with Args_t = {}]’:
>
> ../wtf/fuzzer_hevd.cc:47:41:   required from here
>
> ../wtf/fuzzer_hevd.cc:16:15:   in ‘constexpr’ expansion of ‘fmt::v8::basic_format_string<char>(Format)’
>
> ../wtf/fuzzer_hevd.cc:16:15: error: ‘Format’ is not a constant expression
>
>    16 |     fmt::print(Format, args...);
>
>       |     ~~~~~~~~~~^~~~~~~~~~~~~~~~~
>
> ../wtf/fuzzer_hevd.cc: In instantiation of ‘void Hevd::DebugPrint(const char*, const Args_t& ...) [with Args_t = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]’:
>
> ../wtf/fuzzer_hevd.cc:84:44:   required from here
>
> ../wtf/fuzzer_hevd.cc:16:15:   in ‘constexpr’ expansion of ‘fmt::v8::basic_format_string<char, const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(Format)’
>
> ../wtf/fuzzer_hevd.cc:16:15: error: ‘Format’ is not a constant expression
>
> [2/3] Building CXX object CMakeFiles/wtf.dir/libs/libfuzzer/FuzzerMutate.cpp.o
>
> ninja: build stopped: subcommand failed.
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/0vercl0k/wtf/issues/62>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALIORLJNHFSQEZBZPZGUPDUSF2C7ANCNFSM5KR3LZJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

--- a/src/wtf/bochscpu_backend.cc
+++ b/src/wtf/bochscpu_backend.cc
@@ -13,7 +13,7 @@ constexpr bool BochsHooksLoggingOn = false;
 
 template <typename... Args_t>
 void BochsDebugPrint(const char *Format, const Args_t &...args) {
-  if constexpr (BochsLoggingOn) {
+  if (BochsLoggingOn) {
     fmt::print("bochs: ");
     fmt::print(Format, args...);
   }
@@ -21,7 +21,7 @@ void BochsDebugPrint(const char *Format, const Args_t &...args) {
 
 template <typename... Args_t>
 void BochsHooksDebugPrint(const char *Format, const Args_t &...args) {
-  if constexpr (BochsHooksLoggingOn) {
+  if (BochsHooksLoggingOn) {
     fmt::print("bochshooks: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/bochscpu_backend.cc
+++ b/src/wtf/bochscpu_backend.cc
@@ -15,7 +15,7 @@ template <typename... Args_t>
 void BochsDebugPrint(const char *Format, const Args_t &...args) {
   if constexpr (BochsLoggingOn) {
     fmt::print("bochs: ");
-    fmt::print(Format, args...);
+    fmt::print(fmt::runtime(Format), args...);
   }
 }
 
@@ -23,7 +23,7 @@ template <typename... Args_t>
 void BochsHooksDebugPrint(const char *Format, const Args_t &...args) {
   if constexpr (BochsHooksLoggingOn) {
     fmt::print("bochshooks: ");
-    fmt::print(Format, args...);
+    fmt::print(fmt::runtime(Format), args...);
   }
 }
 

--- a/src/wtf/bochscpu_backend.cc
+++ b/src/wtf/bochscpu_backend.cc
@@ -13,7 +13,7 @@ constexpr bool BochsHooksLoggingOn = false;
 
 template <typename... Args_t>
 void BochsDebugPrint(const char *Format, const Args_t &...args) {
-  if (BochsLoggingOn) {
+  if constexpr (BochsLoggingOn) {
     fmt::print("bochs: ");
     fmt::print(Format, args...);
   }
@@ -21,7 +21,7 @@ void BochsDebugPrint(const char *Format, const Args_t &...args) {
 
 template <typename... Args_t>
 void BochsHooksDebugPrint(const char *Format, const Args_t &...args) {
-  if (BochsHooksLoggingOn) {
+  if constexpr (BochsHooksLoggingOn) {
     fmt::print("bochshooks: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/crash_detection_umode.cc
+++ b/src/wtf/crash_detection_umode.cc
@@ -11,7 +11,7 @@ constexpr bool UCrashDetectionLoggingOn = false;
 
 template <typename... Args_t>
 void CrashDetectionPrint(const char *Format, const Args_t &...args) {
-  if constexpr (UCrashDetectionLoggingOn) {
+  if (UCrashDetectionLoggingOn) {
     fmt::print("ucrash: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/crash_detection_umode.cc
+++ b/src/wtf/crash_detection_umode.cc
@@ -13,7 +13,7 @@ template <typename... Args_t>
 void CrashDetectionPrint(const char *Format, const Args_t &...args) {
   if constexpr (UCrashDetectionLoggingOn) {
     fmt::print("ucrash: ");
-    fmt::print(Format, args...);
+    fmt::print(fmt::runtime(Format), args...);
   }
 }
 

--- a/src/wtf/crash_detection_umode.cc
+++ b/src/wtf/crash_detection_umode.cc
@@ -11,7 +11,7 @@ constexpr bool UCrashDetectionLoggingOn = false;
 
 template <typename... Args_t>
 void CrashDetectionPrint(const char *Format, const Args_t &...args) {
-  if (UCrashDetectionLoggingOn) {
+  if constexpr (UCrashDetectionLoggingOn) {
     fmt::print("ucrash: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/fuzzer_hevd.cc
+++ b/src/wtf/fuzzer_hevd.cc
@@ -11,7 +11,7 @@ constexpr bool LoggingOn = false;
 
 template <typename... Args_t>
 void DebugPrint(const char *Format, const Args_t &...args) {
-  if (LoggingOn) {
+  if constexpr (LoggingOn) {
     fmt::print("Hevd: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/fuzzer_hevd.cc
+++ b/src/wtf/fuzzer_hevd.cc
@@ -13,7 +13,7 @@ template <typename... Args_t>
 void DebugPrint(const char *Format, const Args_t &...args) {
   if constexpr (LoggingOn) {
     fmt::print("Hevd: ");
-    fmt::print(Format, args...);
+    fmt::print(fmt::runtime(Format), args...);
   }
 }
 

--- a/src/wtf/fuzzer_hevd.cc
+++ b/src/wtf/fuzzer_hevd.cc
@@ -11,7 +11,7 @@ constexpr bool LoggingOn = false;
 
 template <typename... Args_t>
 void DebugPrint(const char *Format, const Args_t &...args) {
-  if constexpr (LoggingOn) {
+  if (LoggingOn) {
     fmt::print("Hevd: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/fuzzer_tlv_server.cc
+++ b/src/wtf/fuzzer_tlv_server.cc
@@ -17,7 +17,7 @@ template <typename... Args_t>
 void DebugPrint(const char *Format, const Args_t &...args) {
   if constexpr (LoggingOn) {
     fmt::print("tlv_server: ");
-    fmt::print(Format, args...);
+    fmt::print(fmt::runtime(Format), args...);
   }
 }
 

--- a/src/wtf/fuzzer_tlv_server.cc
+++ b/src/wtf/fuzzer_tlv_server.cc
@@ -15,7 +15,7 @@ constexpr bool LoggingOn = false;
 
 template <typename... Args_t>
 void DebugPrint(const char *Format, const Args_t &...args) {
-  if (LoggingOn) {
+  if constexpr (LoggingOn) {
     fmt::print("tlv_server: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/fuzzer_tlv_server.cc
+++ b/src/wtf/fuzzer_tlv_server.cc
@@ -15,7 +15,7 @@ constexpr bool LoggingOn = false;
 
 template <typename... Args_t>
 void DebugPrint(const char *Format, const Args_t &...args) {
-  if constexpr (LoggingOn) {
+  if (LoggingOn) {
     fmt::print("tlv_server: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/kvm_backend.cc
+++ b/src/wtf/kvm_backend.cc
@@ -26,7 +26,7 @@ template <typename... Args_t>
 void KvmDebugPrint(const char *Format, const Args_t &...args) {
   if constexpr (KvmLoggingOn) {
     fmt::print("kvm: ");
-    fmt::print(Format, args...);
+    fmt::print(fmt::runtime(Format), args...);
   }
 }
 

--- a/src/wtf/kvm_backend.cc
+++ b/src/wtf/kvm_backend.cc
@@ -24,7 +24,7 @@ constexpr bool KvmLoggingOn = false;
 
 template <typename... Args_t>
 void KvmDebugPrint(const char *Format, const Args_t &...args) {
-  if (KvmLoggingOn) {
+  if constexpr (KvmLoggingOn) {
     fmt::print("kvm: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/kvm_backend.cc
+++ b/src/wtf/kvm_backend.cc
@@ -24,7 +24,7 @@ constexpr bool KvmLoggingOn = false;
 
 template <typename... Args_t>
 void KvmDebugPrint(const char *Format, const Args_t &...args) {
-  if constexpr (KvmLoggingOn) {
+  if (KvmLoggingOn) {
     fmt::print("kvm: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/whv_backend.cc
+++ b/src/wtf/whv_backend.cc
@@ -13,7 +13,7 @@ template <typename... Args_t>
 void WhvDebugPrint(const char *Format, const Args_t &...args) {
   if constexpr (WhvLoggingOn) {
     fmt::print("whv: ");
-    fmt::print(Format, args...);
+    fmt::print(fmt::runtime(Format), args...);
   }
 }
 

--- a/src/wtf/whv_backend.cc
+++ b/src/wtf/whv_backend.cc
@@ -11,7 +11,7 @@ constexpr bool WhvLoggingOn = false;
 
 template <typename... Args_t>
 void WhvDebugPrint(const char *Format, const Args_t &...args) {
-  if (WhvLoggingOn) {
+  if constexpr (WhvLoggingOn) {
     fmt::print("whv: ");
     fmt::print(Format, args...);
   }

--- a/src/wtf/whv_backend.cc
+++ b/src/wtf/whv_backend.cc
@@ -11,7 +11,7 @@ constexpr bool WhvLoggingOn = false;
 
 template <typename... Args_t>
 void WhvDebugPrint(const char *Format, const Args_t &...args) {
-  if constexpr (WhvLoggingOn) {
+  if (WhvLoggingOn) {
     fmt::print("whv: ");
     fmt::print(Format, args...);
   }

